### PR TITLE
Introduce global column OID counter

### PR DESCRIPTION
**Pull request by BaurzhanSakhariev**
Taking first 2 commits from https://github.com/crate/crate/pull/14142 to keep it smaller. This is a feature branch, will adjust later if anything is needed.

Naming: I use `column_oid` in Metadata as we [discussed](https://crate.slack.com/archives/CBPLN6WJK/p1678455308356469?thread_ts=1678446209.842239&cid=CBPLN6WJK) internally that PG doesn't expose column oid. We will later introduce global oid for everything but columns will have their own. In FieldMappers (and later in Reference/mapping) I use "oid" since in the context of a column It's clear that it's column oid.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/mapper/GeoShapeFieldMapper.java [hunk 1]
```diff
@@ -187,6 +190,9 @@ public Mapper.Builder<?> parse(String name, Map<String, Object> node, ParserCont
                 } else if ("position".equals(fieldName)) {
                     builder.position(nodeIntegerValue(fieldNode));
                     iterator.remove();
+                } else if ("oid".equals(fieldName)) {
```
[the remark this answers is not in the document]
Comment (BaurzhanSakhariev) [reply server/src/main/java/org/elasticsearch/index/mapper/GeoShapeFieldMapper.java]: Thanks, done in https://github.com/crate/crate/pull/14145/commits/f68914035e3b652e473e809fadbca4787790cbdc

Similar to `TextFieldMapper` - we can call parent `parseField`/`toXContent` methods to handle common fields like position/oid and then let custom code handle specific fields.

File: server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java [hunk 2]
```diff
@@ -141,6 +145,9 @@ protected static boolean parseObjectOrDocumentTypeProperties(String fieldName, O
             if (fieldName.equals("position")) {
                 builder.position(nodeIntegerValue(fieldNode));
                 return true;
+            } else if (fieldName.equals("oid")) {
```
Comment (matriv) [inline server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java:148]: Do we still need this here?

File: server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java [hunk 3]
```diff
@@ -350,6 +359,10 @@ protected void doMerge(final ObjectMapper mergeWith) {
         }
     }
 
+    protected long columnOID() {
```
Comment (matriv) [inline server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java:362]: same

File: server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java [hunk 4]
```diff
@@ -368,6 +381,9 @@ public void toXContent(XContentBuilder builder, Params params, ToXContent custom
         if (position != NOT_TO_BE_POSITIONED) {
             builder.field("position", position);
         }
+        if (columnOID != COLUMN_OID_UNASSIGNED) {
```
Comment (matriv) [inline server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java:384]: same

**Review by matriv — COMMENTED** [review]
Please check one more question

File: server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java (hunk 2 above)
Comment (BaurzhanSakhariev) [reply server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java:148]: ObjectMapper doesn't extend FieldMapper, cannot do smth like  
super.doXContentBody(builder, includeDefaults, params); and handle only specific fields

File: server/src/main/java/org/elasticsearch/index/mapper/GeoShapeFieldMapper.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (BaurzhanSakhariev) [reply server/src/main/java/org/elasticsearch/index/mapper/GeoShapeFieldMapper.java]: I will revert this, GeoShape doesn't support doc values and using common methods also brings doc values handling which makes some tests fail (doc_values starts to appear in the mapping)

File: server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java (hunk 3 above)
Comment (BaurzhanSakhariev) [reply server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java:362]: it's needed it's not a sub class of FieldMapper

**Review by matriv — APPROVED** [review]
LGTM, thx!

**Pull request merged**